test/e2e/helpers.go

```diff
@@ -327,28 +331,98 @@ func getClusterSSHInfo(ctx context.Context, c client.Client, namespace, name str
 	// Collect the info for each instance in a VMSS / MachinePool.
 	machinePools, err := getMachinePoolsInCluster(ctx, c, namespace, name)
 	if err != nil {
-		return nil, err
+		return sshInfo, errors.Wrap(err, "failed to find machine pools in cluster")
+	}
+
+	// make a workload client to access the workload cluster
+	workloadClient, err := getWorkloadClient(ctx, c, namespace, name)
+	if err != nil {
+		return sshInfo, errors.Wrap(err, "failed to get workload client")
 	}
+
 	for i := range machinePools.Items {
 		p := &machinePools.Items[i]
 		cluster, err := util.GetClusterFromMetadata(ctx, c, p.ObjectMeta)
 		if err != nil {
-			return nil, err
+			return sshInfo, errors.Wrap(err, "failed to get cluster from metadata")
+		}
+
+		nodes, err := getReadyNodes(ctx, workloadClient, p.Status.NodeRefs)
+		if err != nil {
+			return sshInfo, errors.Wrap(err, "failed to get ready nodes")
 		}
-		for j := range p.Status.NodeRefs {
-			n := p.Status.NodeRefs[j]
+
+		if p.Spec.Replicas != nil && len(nodes) < int(*p.Spec.Replicas) {
+			message := fmt.Sprintf("machine pool %s/%s expected replicas %d, but only found %d ready nodes", p.Namespace, p.Name, *p.Spec.Replicas, len(nodes))
+			Log(message)
+			return sshInfo, errors.New(message)
+		}
+
+		for _, node := range nodes {
 			sshInfo = append(sshInfo, nodeSSHInfo{
 				Endpoint: cluster.Spec.ControlPlaneEndpoint.Host,
-				Hostname: n.Name,
+				Hostname: node.Name,
 				Port:     e2eConfig.GetVariable(VMSSHPort),
 			})
 		}
-
 	}
 
 	return sshInfo, nil
 }
 
+func getReadyNodes(ctx context.Context, c client.Client, refs []corev1.ObjectReference) ([]corev1.Node, error) {
+	var nodes []corev1.Node
+	for _, ref := range refs {
+		var node corev1.Node
+		if err := c.Get(ctx, client.ObjectKey{
+			Namespace: ref.Namespace,
+			Name:      ref.Name,
+		}, &node); err != nil {
+			if apierrors.IsNotFound(err) {
+				// If 404, continue. Likely the node refs have not caught up to infra providers
+				continue
+			}
+
+			return nodes, err
+		}
+
+		if !noderefutil.IsNodeReady(&node) {
+			Logf("node is not ready and won't be counted for ssh info %s/%s", node.Namespace, node.Name)
+			continue
+		}
+
+		nodes = append(nodes, node)
+	}
+
+	return nodes, nil
+}
+
+func getWorkloadClient(ctx context.Context, c client.Client, namespace, clusterName string) (client.Client, error) {
```

Comment:
is there no helper in the test framework that we can use for this?

Reply:
Good call. Found it via the BootstrapClusterProxy.